File: 0002_add_two_numbers/add_two_numbers.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


def add_two_numbers(l1: ListNode, l2: ListNode) -> ListNode:
    """
    Time: O(m + n)
    Space: O(m + n)
    """
    return add_two_numbers_helper(l1, l2, 0)


def add_two_numbers_helper(l1: ListNode, l2: ListNode,
                           carry: int = 0) -> ListNode:
    if not l1 and not l2 and not carry:
        return None
    total = carry
    if l1:
        total += l1.val
        l1 = l1.next
    if l2:
        total += l2.val
        l2 = l2.next
    carry = 1 if total >= 10 else 0
    new_node = ListNode(total % 10)
    new_node.next = add_two_numbers_helper(l1, l2, carry)
    return new_node

File: 0002_add_two_numbers/add_two_numbers.cpp
#include <iostream>

using namespace std;

class ListNode
{
public:
    int val;
    ListNode *next;
    ListNode(int x)
    {
        val = x;
        next = NULL;
    }
};

ListNode *add_two_numbers_helper(ListNode *l1, ListNode *l2, int carry)
{
    if (!l1 && !l2 && !carry)
        return NULL;
    int temp = (l1 ? l1->val : 0) + (l2 ? l2->val : 0) + carry;
    ListNode *node = new ListNode(temp % 10);
    node->next = add_two_numbers_helper(
        l1 ? l1->next : NULL,
        l2 ? l2->next : NULL,
        temp > 9 ? 1 : 0);
    return node;
}

/**
 * Time: O(m + n)
 * Space: O(m + n)
 */
ListNode *add_two_numbers(ListNode *l1, ListNode *l2)
{
    return add_two_numbers_helper(l1, l2, 0);
}
